tests/python/unittest/test_while_loop.py

```diff
@@ -0,0 +1,969 @@
+            next_hidden, next_states = rnn(in_, states)
+            return [next_hidden], [i + 1] + list(next_states)
+        states = [mx.sym.var("s_" + str(i)) for i in range(n_param)]
+        result = mx.sym.contrib.while_loop(
+                    cond=_cond,
+                    func=_func,
+                    loop_vars=[iter_i] + states,
+                    max_iterations=seq_len
+                )
+        result = mx.sym.Group(result[0] + result[1][1: ])
+        arg_shapes, _, _ = result.infer_shape(
+            data=(seq_len, batch_size, input_dim),
+            s_0=(batch_size, hidden_dim),
+        )
+        rnn_inputs = result.list_inputs()
+        args = {name: _array(arg_shapes[i]) for i, name in enumerate(rnn_inputs)}
+        args_grad = {name: _array(arg_shapes[i]) for i, name in enumerate(rnn_inputs)}
+        e_1 = result.bind(ctx=default_context(),
+            args={name: array.copy() for name, array in args.items()},
+            args_grad={name: array.copy() for name, array in args_grad.items() if name != "i"},
+        )
+        # using unrolled rnn
+        rnn = cell(hidden_dim, prefix='')
+        unroll_outs = []
+        for inputs in mx.sym.split(data, num_outputs=seq_len, axis=0, squeeze_axis=True):
+            h, states = rnn(inputs, states)
+            unroll_outs.append(mx.sym.expand_dims(h, axis=0))
+        unroll_outs = _as_list(mx.sym.concat(*unroll_outs, dim=0))
+        unroll_outs.extend(states)
+        result = mx.sym.Group(unroll_outs)
+        e_2 = result.bind(ctx=default_context(),
+            args={name: array.copy() for name, array in args.items() if name != "i"},
+            args_grad={name: array.copy() for name, array in args_grad.items() if name != "i"},
+        )
+        for case_id in range(5):
+            out_grads = [_array(arr.shape) for arr in e_1.outputs]
+            args = {name: array.copy() for name, array in args.items()}
+            e_1.forward(is_train=True, **args)
+            e_1.backward(out_grads)
+            args = {name: array.copy() for name, array in args.items() if name != "i"}
+            e_2.forward(is_train=True, **args)
+            e_2.backward(out_grads)
+            assert len(e_1.outputs) == len(e_2.outputs)
+            for x, y in zip(e_1.outputs, e_2.outputs):
+                x = x.asnumpy()
+                y = y.asnumpy()
+                assert_almost_equal(x, y, rtol=1e-4, atol=1e-4)
+            grad_keys = list(e_2.grad_dict.keys())
+            e_1_grad = [e_1.grad_dict[x] for x in grad_keys]
+            e_2_grad = [e_2.grad_dict[x] for x in grad_keys]
+            for x, y in zip(e_1_grad, e_2_grad):
+                x = x.asnumpy()
+                y = y.asnumpy()
+                assert_almost_equal(x, y, rtol=1e-4, atol=1e-4)
+
+
+if __name__ == '__main__':
+    # import nose
+    # nose.runmodule()
+    test_while_loop_simple_forward()
+    test_while_loop_for_foreach()
+    test_while_loop_nested()
+    test_while_loop_rnn()
```

Comment:
you don't need to call the test functions explicitly.

Reply:
Oops I forgot to uncomment the lines above. Fixed :-)